### PR TITLE
Developer readme (GSI-696)

### PR DESCRIPTION
**Pull request by TheByronHimes**

### CODE REVIEW

**Review by mephenor — CHANGES_REQUESTED** [review]
Good job on the general dev_guide.
There are some smaller phrasing issues and some points that need a bit more discussion, though.

File: .readme_generation/dev_guide.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+pinning explicit versions of those requirements in what's known as a _lock file_. When
+dependencies are installed for the project, they are installed using that lock file,
+which is stored as [`requirements.txt`](/lock/requirements.txt) in `/lock`.
+Dependencies are shared across all services in the monorepo, so it is not possible, for 
+example, to use hexkit v2 in one service and hexkit v3 in another. 
+There is a templating system in place that allows for cleaner control over dependencies.
+Here's how it works:
+1. In `/lock`, there are several `requirements-*` files:
+  - `requirements-dev-template.in`: Contains *uncapped* (no upper bound) dependencies
+  that are controlled by the Microservice Template Repository and used for development
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md:36]: I think we definitely need a version of that script to keep at least that file in check, but there are probably other things that need to/can be synced too.

File: .readme_generation/dev_guide.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  in all GHGA microservice projects (including monorepos).
+  - `requirements-dev.in`: Contains any other dependencies required for development of
+  this repository specifically.
+  - `requirements-dev.txt`: Lock file containing production *and* development dependencies.
+  - `requirements.txt`: Lock file containing *only* production dependencies.
+
+2. `.pyproject_generation/pyproject_custom.toml` is used to define production-specific
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md:43]: @mephenor Are the application dependencies for all services not aggregated in the pyproject.toml?

File: .readme_generation/dev_guide.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+new images to Docker Hub. To ensure that workflows are only triggered when necessary, the
+[`get_affected_services`](/.github/workflows/get_affected_services.yaml) workflow is used
+to examine git history and only run workflows for services affected by the current branch.
+
+### Docker
+
+Services are deployed in production via docker. In a polyrepo setup, each service maintains
+its own production Dockerfile, but in this monorepo setup there is only one Dockerfile
+for all services. When a release is made, the Dockerfile is used to build a docker image,
+which is then pushed to Docker Hub.
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md:100]: Okay. I'll replace this with TBD and circle back once we've got a concrete solution.

File: .readme_generation/dev_guide.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+## Monorepo Structure
+
+The monorepo houses multiple microservice projects while maintaining isolation between
+each service. Services are located in the [`services`](/services) folder in the root directory.
+| Root-Level Directory | Description |
+|-----------|-------------|
+| [`/.github`](/.github)  | Workflows to be executed on GitHub |
+| [`/.readme_generation`](/.readme_generation) | Template files required to build the root-level README |
+| [`/example_data`](/example_data) | Can be used to populate the application with data for test/dev |
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: Removed

File: .readme_generation/dev_guide.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  that are controlled by the Microservice Template Repository and used for development
+  in all GHGA microservice projects (including monorepos).
+  - `requirements-dev.in`: Contains any other dependencies required for development of
+  this repository specifically.
+  - `requirements-dev.txt`: Lock file containing production *and* development dependencies.
+  - `requirements.txt`: Lock file containing *only* production dependencies.
+
+2. `.pyproject_generation/pyproject_custom.toml` is used to define production-specific
+python dependencies. When changes are made, `scripts/update_pyproject.py` is
+run to update the `pyproject.toml` file at the root level.
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: Reworded

File: .readme_generation/dev_guide.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+- `src`: Application code stored here in a subdirectory labeled with the service abbreviation.
+- `tests_<service_name>`: E.g. `tests_pcs`. All service-specific tests are stored here.
+- Config Files: `config_schema.json`, `dev_config.yaml`, `example_config.yaml`
+- `openapi.yaml`: The OpenAPI specification (only required if using FastAPI).
+- `pyproject.toml`: The service's metadata, used to install the package.
+- `README.md`: Describes the service's purpose, configuration, and design.
+
+### Service Configuration
+
+Configuration is stored in `.yaml` format at the service level. I.e., there is one config
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: Thanks

File: .readme_generation/dev_guide.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+under the `environment` section as <service_name>_CONFIG_YAML. 
+- E.g.: `IRS_CONFIG_YAML: /workspace/services/irs/dev_config.yaml`
+
+### Testing
+
+Testing is performed with `pytest`, which is configured in the pyproject.toml at the repo
+root. At the service level, tests are stored in the folder named `tests_<service_name>`, 
+e.g. `tests_ifrs` for the ifrs project or `tests_pcs` for the pcs project. The tests
+folder lives at the root of the service-specific directory, i.e. 
+`services/<service_name>/tests_<service_name>`.
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: Added

File: .readme_generation/dev_guide.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+Services are deployed in production via docker. In a polyrepo setup, each service maintains
+its own production Dockerfile, but in this monorepo setup there is only one Dockerfile
+for all services. When a release is made, the Dockerfile is used to build a docker image,
+which is then pushed to Docker Hub.
+
+## Service Structure
+
+Services consist of the following high-level parts:
+- `.readme_generation/`: Template components required to build the service-specific README.
+- `scripts/`: Required only if the service uses FastAPI for a REST API.
```
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: Thanks

File: .readme_generation/dev_guide.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mephenor) [reply .readme_generation/dev_guide.md:42]: Yep, but the production specific wording is misleading as those are also needed for testing, etc.

File: .readme_generation/dev_guide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mephenor) [reply .readme_generation/dev_guide.md:35]: I'll add that to my TODO list for the end of the day

File: .readme_generation/dev_guide.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheByronHimes) [reply .readme_generation/dev_guide.md]: I changed it to eliminate ambiguity. Take a look at [13ac5db](https://github.com/ghga-de/file-services-backend/pull/4/commits/13ac5db7b9891361556f42694528409c078fbb80)

**Review by mephenor — APPROVED** [review]

**Pull request merged**